sql: terminate schema change if it violating integrity constraints.

vivekmenezes opened the pull request:
The original code would declare the schema change as unviable
if it hit any error including transient timeouts. This change now
explictly terminates the schema change only if it hits integrity
constraints. Furthermore the schema change is retried for all other
errors with the exception of the "table not found" error.

Whenever an integrity constraint error is hit, the schema change
is first put into reverse, and then backed out. If an error is
seen while backing out, the process is stopped to be continued
later via the asynchronous schema change manager.

fixes #6333 #6432

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6552)

<!-- Reviewable:end -->

vivekmenezes committed:
sql: terminate schema change if it violating integrity constraints.

The original code would declare the schema change as unviable
if it hit any error including transient timeouts. This change now
explictly terminates the schema change only if it hits integrity
constraints. Furthermore the schema change is retried for all other
errors with the exception of the "table not found" error.

Whenever an integrity constraint error is hit, the schema change
is first put into reverse, and then backed out. If an error is
seen while backing out, the process is stopped to be continued
later via the asynchronous schema change manager.

tamird commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[sql/errors.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aByPFhKjtVhhvkkK:-KH5aByPFhKjtVhhvkkL:1527469067) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L36)):*
:+1: neat

---

*[sql/errors.go, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aG3r1xFHAyOdD4bt:-KH5aG3r1xFHAyOdD4bu:-1593029036) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L119)):*
nit: s/column/columnName/?

---

*[sql/errors.go, line 123 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aRtTyCvf0ktNtOU3:-KH5aRtU3ZRsE4rt394L:-1153297394) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L123)):*
pg gives a better error: `null value in column "id" violates not-null constraint`

---

*[sql/schema_changer.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bswaxfTwTFJ_y1tZ:-KH5bswbi-ulf_VL7Xj0:-1707864528) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L167)):*
s/schemaChangeRetryError/isSchemaChangeRetryError/g

---

*[sql/schema_changer.go, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5ag-LJO5e3EVshK43:-KH5ag-LJO5e3EVshK44:1911583169) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L168)):*
nit: i like the type-switch style better like you have in isIntegrityConstraintError

```
switch err.(type) {
  case *roachpb.DescriptorNotFoundError:
    return false
  default:
    return isIntegrityConstraitError(err)
}
```

---

*[sql/schema_changer.go, line 242 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5azW1NyFwEvCvAx1w:-KH5azW22fZZ-QZ-zYu1:229717506) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L242)):*
s/fail/failed/

---

*[sql/schema_changer.go, line 244 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5b0MW-8TdviciQx3j:-KH5b0MW-8TdviciQx3k:917373085) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L244)):*
"accomodate" doesn't feel like the right word here. Perhaps "that may be resolved by retrying"?

---

*[sql/schema_changer.go, line 247 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bDSCqrTXMzsYfvVL:-KH5bDSCqrTXMzsYfvVM:-898350852) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L247)):*
s/constraint/constraint violation/

---

*[sql/schema_changer.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bKSkmEtMEIzIu8xu:-KH5bKSkmEtMEIzIu8xv:-536791259) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L249)):*
sounds scary. does this get logged?

---

*[sql/schema_changer.go, line 255 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bOQMVJRjjVefKIbf:-KH5bOQMVJRjjVefKIbg:199974796) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L255)):*
scope this more narrowly.

---

*[sql/schema_changer.go, line 409 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bg5BrHowAmcGOotq:-KH5bg5BrHowAmcGOotr:1496110825) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L409)):*
put this in the caller and log the actual error - it's not always an integrity constraint violation (as evidenced by the existence of `schemaChangeRetryError`)

---

*[sql/schema_changer_test.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cMlMyGSDytOVw3NC:-KH5cMlMyGSDytOVw3ND:2064788808) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L726)):*
nit: why not composite literal?

---

*[sql/schema_changer_test.go, line 730 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cJ_SO0lb2emMwqk9:-KH5cJ_SO0lb2emMwqkA:-1728228851) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L730)):*
is this thread safe?

---

*[sql/schema_changer_test.go, line 744 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cfJPFFp4-Pl14SpQ:-KH5cfJPFFp4-Pl14SpR:-1874540878) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L744)):*
nit: `k` appears not to be used. can you use `CREATE TABLE t.test (v INT)`?

---

*[sql/schema_changer_test.go, line 765 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dPhR3fPjWVW5Ohzx:-KH5dPhR3fPjWVW5Ohzy:-1336367602) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L765)):*
what if attempts is 3 or 1? be more specific please.

---

*[sql/schema_changer_test.go, line 808 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5d6gARNBEg5b5Gmte:-KH5d6gARNBEg5b5Gmtf:-1037169502) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L808)):*
nit: you can scope this more narrowly with
```
defer func (backFillChunkSize int) {
  csql.IndexBackfillChunkSize = backfillChunkSize
}(csql.IndexBackfillChunkSize)
```

really don't feel strongly about this one.

---

*[sql/schema_changer_test.go, line 814 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dKXZP84pFA-C3OmI:-KH5dKXZP84pFA-C3OmJ:-1728228851) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L814)):*
is this thread safe?

---

*[sql/schema_changer_test.go, line 833 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dU_i_BP_hXJBWWup:-KH5dU_i_BP_hXJBWWuq:-1057368652) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L833)):*
mention the expected outcome

---

*[sql/schema_changer_test.go, line 837 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dZJcplh4QGFT399A:-KH5dZJcplh4QGFT399B:383486821) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L837)):*
"not found" is way too broad

---

*[sql/schema_changer_test.go, line 860 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dhYpLqd0N1wpefkz:-KH5dhYpLqd0N1wpefl-:1581310202) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L860)):*
mention the expected outcome.

---

*[sql/schema_changer_test.go, line 865 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dki2PthG_JkCU3EW:-KH5dki2PthG_JkCU3EX:1581310202) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L865)):*
mention the expected outcome.

---

*[sql/schema_changer_test.go, line 868 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dqoED-nJGRymnAUY:-KH5dqoED-nJGRymnAUZ:-502165291) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L868)):*
can you also test the reversal to make sure this gets cleaned up? perhaps this is tested elsewhere? if so, where?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> #### sql: terminate schema change if it violating integrity constraints\.
> 
> The original code would declare the schema change as unviable
> 
> if it hit any error including transient timeouts. This change now
> 
> explictly terminates the schema change only if it hits integrity
> 
> constraints. Furthermore the schema change is retried for all other
> 
> errors with the exception of the "table not found" error.
> 
> 
> 
> Whenever an integrity constraint error is hit, the schema change
> 
> is first put into reverse, and then backed out. If an error is
> 
> seen while backing out, the process is stopped to be continued
> 
> later via the asynchronous schema change manager.
> 
> 
> 
> fixes #6333 #6432
> 
> 
> </details>


---

Review status: all files reviewed at latest revision, 28 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLW-1kqXT0u40NI0A0:-KHLW-1kqXT0u40NI0A1:328259357) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/backfill.go#L407)):*
> ```Go
> // adjust it dynamically. Also add in a sleep after every chunk is processed,
> // to slow down the backfill and not have it interfere with OLTP commands.
> var IndexBackfillChunkSize = 100
> ```

can this really be updated without a race?

---

*[sql/errors.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aByPFhKjtVhhvkkK:-KHLcO4gTODuJOVazgcQ:1565160846) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> :+1: neat</details>
neatto indeed. Mind asserting this error code in your new test (I think you'll have to go through lib/pq)? And also asserting it in a LogicTest (I've added a `pgcode` clause to the logic test language a while ago).

---

*[sql/executor.go, line 217 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLWc0HczP1UFZWMlNW:-KHLWc0ILWys-2zGsKcW:-1731279736) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/executor.go#L217)):*
> ```Go
> 	// SchemaChangersStartBackfillNotification is called before applying the
> 	// backfill for a schema change operation.
> 	SchemaChangersStartBackfillNotification func() error
> ```

can you please comment on what it means for this function to return an error

---

*[sql/schema_changer.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bswaxfTwTFJ_y1tZ:-KHLWqeDRNq5p6HE1XFH:354258735) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/schemaChangeRetryError/isSchemaChangeRetryError/g</details>
how about `canRetry` ?

---

*[sql/schema_changer.go, line 642 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLZ8NC9QaOKQFQ3pk7:-KHLZ8NC9QaOKQFQ3pk8:-1867583196) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L642)):*
> ```Go
> 							} else {
> 								// We don't need to act on integrity
> 								// constraints violations because exec() is
> ```

nit: s/is capable of purging/purges

---

*[sql/schema_changer_test.go, line 729 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLa8z6RLgAptxfQzPB:-KHLa8z6RLgAptxfQzPC:1747460171) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L729)):*
> ```Go
> 	attempts := 0
> 	execKnobs.SchemaChangersStartBackfillNotification =
> 		func() error {
> ```

nit: move the error injection to just before the `CREATE INDEX` statement. I.e. make execKnobs.SchemaChangeSBN a thread-safe shim that calls another function, and set the inner function later on.

---

*[sql/schema_changer_test.go, line 803 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLbPJErnFpbC68Wpib:-KHLbPJErnFpbC68Wpic:-1115609373) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L803)):*
> ```Go
> 	}
> 
> 	// Now recreate index where a constraint violation triggers a reverse of
> ```

mind splitting the rest of this into a new test?

---

*[sql/schema_changer_test.go, line 812 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLbh15RG8yPl55xuqk:-KHLbh15RG8yPl55xuql:1142923944) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L812)):*
> ```Go
> 		csql.IndexBackfillChunkSize = backfillChunkSize
> 	}()
> 	csql.IndexBackfillChunkSize = 1
> ```

how about instead of reducing the BackfillChunkSize, we generate enough rows to span multiple chunks (so we used the chunk size constant in the batch generation above), and we make a duplicate in the 2nd chunk.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-:-KHLcrotTgKlbMiVX0qK:1417088721)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 7 files reviewed at latest revision, 27 unresolved discussions, some commit checks pending.

---

*[sql/errors.go, line 119 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aG3r1xFHAyOdD4bt:-KHL2Xf8LJtYDUfQ6Wwr:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L119)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: s/column/columnName/?</details>
done

---

*[sql/errors.go, line 123 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aRtTyCvf0ktNtOU3:-KHL2xDLRWxKSztlMzdZ:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L123)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> pg gives a better error: `null value in column "id" violates not-null constraint`</details>
done

---

*[sql/schema_changer.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bswaxfTwTFJ_y1tZ:-KHL35xa5y3RODJshQZv:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L167)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/schemaChangeRetryError/isSchemaChangeRetryError/g</details>
done

---

*[sql/schema_changer.go, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5ag-LJO5e3EVshK43:-KHL37BeLpuG7cowMh6i:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: i like the type-switch style better like you have in isIntegrityConstraintError
> 
> ```
> switch err.(type) {
>   case *roachpb.DescriptorNotFoundError:
>     return false
>   default:
>     return isIntegrityConstraitError(err)
> }
> ```</details>
done

---

*[sql/schema_changer.go, line 242 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5azW1NyFwEvCvAx1w:-KHL3SWOyHn8rlodyXc7:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L242)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/fail/failed/</details>
done

---

*[sql/schema_changer.go, line 244 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5b0MW-8TdviciQx3j:-KHL3T95LvksBUtFSV8-:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "accomodate" doesn't feel like the right word here. Perhaps "that may be resolved by retrying"?</details>
done

---

*[sql/schema_changer.go, line 247 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bDSCqrTXMzsYfvVL:-KHL3bEFCImI-ag624-I:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L247)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/constraint/constraint violation/</details>
done

---

*[sql/schema_changer.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bKSkmEtMEIzIu8xu:-KHL3cv6o2lesL80Fqz3:-382938660) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L249)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> sounds scary. does this get logged?</details>
It gets logged in session.go. Its actually not scary because the async path takes care of node failures and we do test that a good state is left behind 

---

*[sql/schema_changer.go, line 255 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bOQMVJRjjVefKIbf:-KHL5gG0bneq1FhKed4E:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L255)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> scope this more narrowly.</details>
done

---

*[sql/schema_changer.go, line 409 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bg5BrHowAmcGOotq:-KHL5UBT9CswgQ9Y8mqt:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L409)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> put this in the caller and log the actual error - it's not always an integrity constraint violation (as evidenced by the existence of `schemaChangeRetryError`)</details>
done

---

*[sql/schema_changer_test.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cMlMyGSDytOVw3NC:-KHL7brnO2n4xWZHH0Mj:-584183475) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L726)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: why not composite literal?</details>
to allow us to reassign SchemaChangersStartBackfillNotification later

---

*[sql/schema_changer_test.go, line 730 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cJ_SO0lb2emMwqk9:-KHL5qsz46bPH2bFJGNU:-1763578873) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L730)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this thread safe?</details>
it doesn't have to be thread safe because it's called in a specific order.

---

*[sql/schema_changer_test.go, line 744 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cfJPFFp4-Pl14SpQ:-KHL8-TsKgMxTUovb6t4:1963367944) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L744)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: `k` appears not to be used. can you use `CREATE TABLE t.test (v INT)`?</details>
we need k to maintain a primary key on the table, besides it's a kv table

---

*[sql/schema_changer_test.go, line 765 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dPhR3fPjWVW5Ohzx:-KHLIvV1XzoLNXRn7LSu:-379376322) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L765)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what if attempts is 3 or 1? be more specific please.</details>
it now logs the number of attempts

---

*[sql/schema_changer_test.go, line 808 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5d6gARNBEg5b5Gmte:-KHLJMvTONkTnVr6HNV8:-837966666) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L808)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: you can scope this more narrowly with
> ```
> defer func (backFillChunkSize int) {
>   csql.IndexBackfillChunkSize = backfillChunkSize
> }(csql.IndexBackfillChunkSize)
> ```
> 
> really don't feel strongly about this one.</details>
nice. Done

---

*[sql/schema_changer_test.go, line 814 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dKXZP84pFA-C3OmI:-KHLJPJV5PhiMyMNFA_a:-524025497) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L814)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this thread safe?</details>
again, the retries occur in a specific order and there isn't a need for any locking.

---

*[sql/schema_changer_test.go, line 833 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dU_i_BP_hXJBWWup:-KHLKbvD_kH5ZX--NjS4:679869031) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L833)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> mention the expected outcome</details>
is that better?

---

*[sql/schema_changer_test.go, line 837 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dZJcplh4QGFT399A:-KHLL14bFIw9jE3S38YA:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L837)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "not found" is way too broad</details>
fixed

---

*[sql/schema_changer_test.go, line 860 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dhYpLqd0N1wpefkz:-KHLUYks9wod87Xj0ytG:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L860)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> mention the expected outcome.</details>
done

---

*[sql/schema_changer_test.go, line 865 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dki2PthG_JkCU3EW:-KHLUXdUhfXpE22Bqrfr:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L865)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> mention the expected outcome.</details>
done

---

*[sql/schema_changer_test.go, line 868 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dqoED-nJGRymnAUY:-KHLnQzeVachuhYRsNN-:-616844176) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L868)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you also test the reversal to make sure this gets cleaned up? perhaps this is tested elsewhere? if so, where?</details>
Done. I had to export disableAsyncSchemaChangeExec to allow us to turn it on and then off later when we require the async path to run and clean up the garbage left behind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 7 files reviewed at latest revision, 27 unresolved discussions, some commit checks pending.

---

*[sql/schema_changer_test.go, line 868 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dqoED-nJGRymnAUY:-KHLqMhjWkXax9gldpGi:1469660545) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L868)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done. I had to export disableAsyncSchemaChangeExec to allow us to turn it on and then off later when we require the async path to run and clean up the garbage left behind.</details>
I think probably this whole file should move to the `sql` package (which you can now do because of Radu's work - you can start a server without depending on the server package)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[sql/schema_changer.go, line 179 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLsb43AABW7LBbcpsv:-KHLsb44durqDLrKC3Ti:689912012) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f150eaeb0f5270726974fa9d41274f412cb35a/sql/schema_changer.go#L179)):*
> ```Go
> // called before the backfill starts; it can be nil.
> func (sc SchemaChanger) exec(startBackfillNotification func() error) error {
> 	log.Info("excuting schema change")
> ```

why'd this sneak in?

---

*[sql/schema_changer.go, line 245 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLrIv9FJOmwiOVAr6G:-KHLrIvAWQGHp9dv9XYY:-1725805133) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f150eaeb0f5270726974fa9d41274f412cb35a/sql/schema_changer.go#L245)):*
> ```Go
> 	err = sc.runStateMachineAndBackfill(&lease, startBackfillNotification)
> 
> 	// Purge the mutations if the application of the mutations failed due to a
> ```

s/a/an/

---

*[sql/schema_changer.go, line 249 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLsv9igLE11jMOxjJo:-KHLsv9jXL2LiCpWVqii:-1250206219) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f150eaeb0f5270726974fa9d41274f412cb35a/sql/schema_changer.go#L249)):*
> ```Go
> 	// that are resolved by retrying the backfill.
> 	if isIntegrityConstraintError(err) {
> 		log.Warning("reversing schema change due to irrecoverable error %s", err)
> ```

s/error/error:/

---

*[sql/schema_changer_test.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cMlMyGSDytOVw3NC:-KHLs4KdzFjXJJc0WyyQ:1357881739) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L726)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> to allow us to reassign SchemaChangersStartBackfillNotification later</details>
a composite literal also allows you to do that.

---

*[sql/schema_changer_test.go, line 744 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cfJPFFp4-Pl14SpQ:-KHLs05TPwkLcDz5BRvO:1944975054) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L744)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> we need k to maintain a primary key on the table, besides it's a kv table</details>
why do you need a primary key?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 3 of 7 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLW-1kqXT0u40NI0A0:-KHLqPJdHJ2Sb4ay14Zs:-254565092) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/backfill.go#L407)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can this really be updated without a race?</details>
well not in theory, but this is a test and in this test there is no race.

---

*[sql/errors.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5aByPFhKjtVhhvkkK:-KHLqrwbMFd5C0sj0_lk:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/errors.go#L36)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> neatto indeed. Mind asserting this error code in your new test (I think you'll have to go through lib/pq)? And also asserting it in a LogicTest (I've added a `pgcode` clause to the logic test language a while ago).</details>
done

---

*[sql/executor.go, line 217 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLWc0HczP1UFZWMlNW:-KHLsi0Bs8ERvnJbNqbx:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/executor.go#L217)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you please comment on what it means for this function to return an error</details>
done

---

*[sql/schema_changer.go, line 167 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5bswaxfTwTFJ_y1tZ:-KHLrihBiHlXC_9XSC_u:1839303401) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L167)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> done</details>
canRetry works if it is tied to SchemaChanger.

---

*[sql/schema_changer.go, line 642 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLZ8NC9QaOKQFQ3pk7:-KHLryEBXkXnZ-NgcMy5:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer.go#L642)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: s/is capable of purging/purges</details>
done

---

*[sql/schema_changer.go, line 179 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLsb43AABW7LBbcpsv:-KHLxbeG7IIPr8tXtLFT:1019960892) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f150eaeb0f5270726974fa9d41274f412cb35a/sql/schema_changer.go#L179)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why'd this sneak in?</details>
sneaked out

---

*[sql/schema_changer.go, line 245 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLrIv9FJOmwiOVAr6G:-KHLxlElVNajw34Tag3Z:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f150eaeb0f5270726974fa9d41274f412cb35a/sql/schema_changer.go#L245)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/a/an/</details>
done

---

*[sql/schema_changer.go, line 249 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLsv9igLE11jMOxjJo:-KHLyK7a87A6WxNNxKAY:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/98f150eaeb0f5270726974fa9d41274f412cb35a/sql/schema_changer.go#L249)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/error/error:/</details>
done

---

*[sql/schema_changer_test.go, line 726 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cMlMyGSDytOVw3NC:-KHM-DVaVO_rjA3KH44U:-341286575) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L726)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> a composite literal also allows you to do that.</details>
all fixed!

---

*[sql/schema_changer_test.go, line 729 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLa8z6RLgAptxfQzPB:-KHM-tfGYsIF64N4qOyj:-489179633) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L729)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: move the error injection to just before the `CREATE INDEX` statement. I.e. make execKnobs.SchemaChangeSBN a thread-safe shim that calls another function, and set the inner function later on.</details>
Not sure what you mean. Does your comment still make sense given all the other changes I've made here?

---

*[sql/schema_changer_test.go, line 744 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cfJPFFp4-Pl14SpQ:-KHM-I4RWmFgtOT6nDSb:2097980735) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L744)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why do you need a primary key?</details>
We add a duplicate value for v to violate the UNIQUE integrity constraint, and every table needs a primary key so we do need another column which is k. Since I've split this into two tests now technically I only need k:v in the second test, but it's just nicer to be consistent.

---

*[sql/schema_changer_test.go, line 803 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLbPJErnFpbC68Wpib:-KHLx2rB-T5kKKlZ8082:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L803)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind splitting the rest of this into a new test?</details>
done

---

*[sql/schema_changer_test.go, line 812 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHLbh15RG8yPl55xuqk:-KHLxB1f2ZLyjMITuHRx:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L812)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about instead of reducing the BackfillChunkSize, we generate enough rows to span multiple chunks (so we used the chunk size constant in the batch generation above), and we make a duplicate in the 2nd chunk.</details>
done

---

*[sql/schema_changer_test.go, line 868 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5dqoED-nJGRymnAUY:-KHLxPtcc8iOCkArxg0L:208543424) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L868)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think probably this whole file should move to the `sql` package (which you can now do because of Radu's work - you can start a server without depending on the server package)</details>
good idea. I'll send that out as a separate PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[sql/schema_changer_test.go, line 744 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KH5cfJPFFp4-Pl14SpQ:-KHMCRcgOfzxBJDl8GDa:990763155) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff0f9707b6bc493aafee22b9538a7c4c137bba87/sql/schema_changer_test.go#L744)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> We add a duplicate value for v to violate the UNIQUE integrity constraint, and every table needs a primary key so we do need another column which is k. Since I've split this into two tests now technically I only need k:v in the second test, but it's just nicer to be consistent.</details>
We do support tables without primary keys, but ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
This testing is failing make testrace. I'm going to add a testing knob to fix that. Coming up

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> :lgtm:</details>


---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-:-KHMHHvArvGcSXL4X8i7:538297793)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
This is ready for another look. I've also filed away https://github.com/cockroachdb/cockroach/issues/6596 to cleanup the schema change testing knobs. I tried doing it in this change, but it kept growing.

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> This testing is failing make testrace. I'm going to add a testing knob to fix that. Coming up</details>


---

Review status: 4 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-:-KHNCqGDEd1LyWkv9jNa:1702540769)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I'm going to merge this. There are obvious improvements I can make to the testing knobs. I'm going to do a follow up PR to split away the schema change testing knobs from the executor testing knobs and change all the global testing knobs to context testing knobs.

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> This is ready for another look. I've also filed away https://github.com/cockroachdb/cockroach/issues/6596 to cleanup the schema change testing knobs. I tried doing it in this change, but it kept growing.</details>


---

Review status: 4 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-:-KHQ8hNvO9WUWWZdboIX:1056427258)*
<!-- Sent from Reviewable.io -->

tamird commented:
Can you wait until the last changes have been reviewed?

<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I'm going to merge this. There are obvious improvements I can make to the testing knobs. I'm going to do a follow up PR to split away the schema change testing knobs from the executor testing knobs and change all the global testing knobs to context testing knobs.</details>


---

Review status: 4 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-:-KHQAJYfi80TF9FAvEwj:630369355)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
sure

On Tue, May 10, 2016 at 10:59 AM Tamir Duberstein <notifications@github.com>
wrote:

> Can you wait until the last changes have been reviewed?
>
> *Previously, vivekmenezes wrote…*
>
> I'm going to merge this. There are obvious improvements I can make to the
> testing knobs. I'm going to do a follow up PR to split away the schema
> change testing knobs from the executor testing knobs and change all the
> global testing knobs to context testing knobs.
>
> ------------------------------
>
> Review status: 4 of 8 files reviewed at latest revision, 10 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-:-KHQAJYfi80TF9FAvEwj:630369355>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6552#issuecomment-218184653>
>

tamird commented:
Reviewed 3 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[server/server.go, line 390 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQBATeVujX5dnG2Uvk:-KHQBATeVujX5dnG2Uvl:2128814548) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/server/server.go#L390)):*
> ```Go
> 	sqlTestKnobs := &sql.ExecutorTestingKnobs{}
> 	if s.ctx.TestingKnobs.SQLExecutor != nil {
> 		sqlTestKnobs = s.ctx.TestingKnobs.SQLExecutor.(*sql.ExecutorTestingKnobs)
> ```

what's with this type assertion? is SQLExecutor an interface?

---

*[sql/executor.go, line 222 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAvniE1lEhw6Otf2a:-KHQAvniE1lEhw6Otf2b:-276358514) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/executor.go#L222)):*
> ```Go
> 	// AsyncSchemaChangersExecNotification is a function called before running
> 	// a schema change asynchronously. Returning an error will prevent the
> 	// asynchronously execution path from running.
> ```

s/asynchronously/asynchronous/

---

*[sql/schema_changer_test.go, line 798 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAhJHZaqnS2xHxTVL:-KHQAhJHZaqnS2xHxTVM:-791929451) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/schema_changer_test.go#L798)):*
> ```Go
> 	defer csql.TestSpeedupAsyncSchemaChanges()()
> 
> 	// mu protects disableAsyncSchemaChanges.
> ```

use atomics?

---

*[sql/schema_changer_test.go, line 819 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAVZjqahQQVVFSxVG:-KHQAVZk2NTsSNcUC5HY:-878408524) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/schema_changer_test.go#L819)):*
> ```Go
> 			defer mu.Unlock()
> 			if disableAsyncSchemaChanges {
> 				return errors.New("don't process schema change")
> ```

can you give this the same tense as the other error immediately above? "async schema changes are disabled".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 390 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQBATeVujX5dnG2Uvk:-KHQDE1jVOg_ge535agk:-999334156) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/server/server.go#L390)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's with this type assertion? is SQLExecutor an interface?</details>
see base/testing_knobs.go. I plan to add a new SQLSchemaChanger knob and move the schema changer knobs in there.

---

*[sql/executor.go, line 222 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAvniE1lEhw6Otf2a:-KHQDiUFrY9KM3SI3DVo:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/executor.go#L222)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/asynchronously/asynchronous/</details>
done

---

*[sql/schema_changer_test.go, line 798 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAhJHZaqnS2xHxTVL:-KHQEIzShRwCvGqtfxN_:942151733) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/schema_changer_test.go#L798)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> use atomics?</details>
there is no atomic bool

---

*[sql/schema_changer_test.go, line 819 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAVZjqahQQVVFSxVG:-KHQE2tgwhptHmofoWoC:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/schema_changer_test.go#L819)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you give this the same tense as the other error immediately above? "async schema changes are disabled".</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[sql/schema_changer_test.go, line 798 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAhJHZaqnS2xHxTVL:-KHQFSUKoF0vLvz9oEUO:-110947152) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/schema_changer_test.go#L798)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> there is no atomic bool</details>
seems like an int would suffice, or an atomic.Value?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 7 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[sql/schema_changer_test.go, line 798 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552#-KHQAhJHZaqnS2xHxTVL:-KHQLZfGuw2XuFHKYVcF:519048634) ([raw file](https://github.com/cockroachdb/cockroach/blob/79afe8d84e7074dd5a50756202b7dd5f086bbf9e/sql/schema_changer_test.go#L798)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like an int would suffice, or an atomic.Value?</details>
It now uses atomic with an int32

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6552)*
<!-- Sent from Reviewable.io -->
